### PR TITLE
feat: Implement PRC-4: Authenticated Data with Associated Data

### PR DESCRIPTION
**Pull request by explodingcamera**
Specification: https://pog.network/specification/PIPs/04-ADAD/
This PR doesn't actually implement ADAD for our datastructures, this will happen in a seperate PR.

### CODE REVIEW

File: champ/lib/encoding/adad.rs [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+use thiserror::Error;
+use tokio::io::{AsyncRead, AsyncReadExt};
+
+#[derive(Error, Debug)]
+pub enum ADADError {
+    #[error("unknown error")]
+    Unknown,
+    #[error("failed to read associated data length")]
+    VarIntReadError(#[from] unsigned_varint::io::ReadError),
+    #[error("failed to read associated data length")]
```
Comment (Alyks1) [inline champ/lib/encoding/adad.rs:14]: Maybe change this msg to distinguish between the one before?

File: champ/lib/encoding/adad.rs [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    let codec = unsigned_varint::io::read_usize(&mut reader)?;
+    let mut authenticated_data = vec![];
+    reader.read_to_end(&mut authenticated_data)?;
+    Ok((authenticated_data, codec))
+}
+
+/// Reads associated data from a reader
+pub fn read_associated_data<T: std::io::Read>(mut reader: T) -> Result<(Vec<u8>, usize), ADADError> {
+    let length = unsigned_varint::io::read_usize(&mut reader)?;
+    let codec = unsigned_varint::io::read_usize(&mut reader)?;
```
Comment (Alyks1) [inline champ/lib/encoding/adad.rs:81]: For clarity it makes sense, but maybe we could refactor these?

File: champ/lib/encoding/adad.rs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    const VALID_DATA: [(&[u8], &[u8]); 5] = [
+        ("test_header".as_bytes(), "test_data".as_bytes()),
+        ("".as_bytes(), "".as_bytes()),
+        ("".as_bytes(), "test_data".as_bytes()),
+        ("test_header".as_bytes(), "".as_bytes()),
+        ("test_header".as_bytes(), "long_test_data".as_bytes()),
+    ];
+
+    #[test]
+    fn test() {
```
Comment (Alyks1) [inline champ/lib/encoding/adad.rs:177]: We might need tests for the proto codec :P

File: champ/lib/encoding/adad.rs (hunk 1 above)
Comment (explodingcamera) [reply champ/lib/encoding/adad.rs:14]: oh rip that was suppoesed to be a transparent error

**Pull request merged**